### ISSUE TITLE
[vm/ffi] Allow extension types that implements `NativeType`s in type arguments

### ISSUE THREAD

**Issue by halildurmus** [dart-lang/sdk#54944]
Currently, we cannot use extension types that implement `NativeType`s as type arguments in FFI. Given that extension types do not exist at runtime, I cannot find any reason why this should not be allowed.

Here's an example:
```dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

extension type const ExtInt32(Int32 _) implements Int32 {}

base class MyStruct extends Struct {
  @Int32()
  external int a;

  @ExtInt32()
  // ERROR: Fields of type 'int' in a subclass of 'Struct' must have an
  // annotation indicating the native type.
  external int b;
}

extension type const ExtMyStruct(MyStruct _) implements MyStruct {}

@Native<ExtInt32 Function(ExtInt32)>() // no error here
external int fn(int p);

void main() {
  // ERROR: The type arguments to 'AllocatorAlloc.call' must be known at compile
  // time, so they can't be type parameters.
  calloc<ExtInt32>();
  calloc<ExtMyStruct>();

  final p = calloc<Int32>().cast<ExtInt32>()..value = 42; // works
  print(p.value); // prints 42

  final s = calloc<MyStruct>();
  // ERROR: The type arguments to 'ref' must be known at compile time, so they
  // can't be type parameters.
  s.cast<ExtMyStruct>().ref;

  // ERROR: The type arguments to 'Struct.create' must be known at compile time,
  // so they can't be type parameters.
  Struct.create<ExtMyStruct>();

  // ERROR: The type arguments to 'sizeOf' must be known at compile time, so
  // they can't be type parameters.
  sizeOf<ExtInt32>();
  sizeOf<ExtMyStruct>();

  // ERROR: The type 'Int32 Function(Int32)' must be a subtype of
  // 'ExtInt32 Function(ExtInt32)' for 'Native.addressOf'.
  Native.addressOf<NativeFunction<ExtInt32 Function(ExtInt32)>>(fn);

  final p2 =
      Pointer<NativeFunction<ExtInt32 Function(ExtInt32)>>.fromAddress(0);
  // ERROR: Can't invoke 'asFunction' because the function signature
  // 'NativeFunction<Int32 Function(ExtInt32)>' for the pointer isn't a valid
  // C function signature.
  p2.asFunction<int Function(int)>();
}
```

cc @dcharkes

**Comment by dcharkes** [comment]
@halildurmus Can you detail a bit more about the use case where this is useful?

**Comment by halildurmus** [comment]
**TL;DR:** I would like to use extension types like `typedef` for native types in function signatures to provide users with clearer information and thus enable easier consumption of Win32 APIs within Dart.

I'm currently working on `v6` of the `package:win32` with a focus on enhancing the user experience when using the Win32 APIs.

Let's take a specific example with the [CoGetApartmentType](https://learn.microsoft.com/windows/win32/api/combaseapi/nf-combaseapi-cogetapartmenttype) function, which is defined in C as follows:
```c
HRESULT CoGetApartmentType(APTTYPE *pAptType, APTTYPEQUALIFIER *pAptQualifier);

typedef enum _APTTYPE {
  APTTYPE_CURRENT = -1,
  APTTYPE_STA = 0,
  APTTYPE_MTA = 1,
  APTTYPE_NA = 2,
  APTTYPE_MAINSTA = 3
} APTTYPE;

typedef enum _APTTYPEQUALIFIER {
  APTTYPEQUALIFIER_NONE = 0,
  APTTYPEQUALIFIER_IMPLICIT_MTA = 1,
  APTTYPEQUALIFIER_NA_ON_MTA = 2,
  APTTYPEQUALIFIER_NA_ON_STA = 3,
  APTTYPEQUALIFIER_NA_ON_IMPLICIT_MTA = 4,
  APTTYPEQUALIFIER_NA_ON_MAINSTA = 5,
  APTTYPEQUALIFIER_APPLICATION_STA = 6,
  APTTYPEQUALIFIER_RESERVED_1 = 7
} APTTYPEQUALIFIER;
```

In `package:win32`, this function is projected into Dart as follows:
```dart
int CoGetApartmentType(Pointer<Int32> pAptType, Pointer<Int32> pAptQualifier) =>
    _CoGetApartmentType(pAptType, pAptQualifier);

final _CoGetApartmentType = _ole32.lookupFunction<
    Int32 Function(Pointer<Int32> pAptType, Pointer<Int32> pAptQualifier),
    int Function(Pointer<Int32> pAptType,
        Pointer<Int32> pAptQualifier)>('CoGetApartmentType');
```

However, it's not immediately clear to the user what the `pAptType` and `pAptQualifier` parameters represent. Users need to refer back to the Microsoft documentation to understand that they are meant to be `APTTYPE` and `APTTYPEQUALIFIER` enums. Even if I chose to represent these enums as `typedef`s, they would still see `Pointer<Int32>` type when they hover over the function (issue #39332).

(I should also note that currently, the projected APIs in `package:win32` are manually defined in a [JSON file](https://github.com/dart-windows/win32/blob/main/tool/win32gen/data/win32_functions.json#L620-L623), including the [C function definition as documentation](https://github.com/dart-windows/win32/blob/a78ff108fbf54927b0334c1bab8f15526c24cbf8/lib/src/win32/ole32.g.dart#L148-L151), which is extracted from the [Microsoft documentation website](https://learn.microsoft.com/windows/win32/api/combaseapi/nf-combaseapi-cogetapartmenttype#syntax). However, in `v6`, instead of manually defining the projected APIs in a JSON file, I plan to project the entire Win32 API set (excluding a few ancient APIs) using Microsoft's metadata. Consequently, users won't see these manually added C definitions when they hover over the function in their IDE. Instead, they'll just see the function signature `int CoGetApartmentType(Pointer<Int32> pAptType, Pointer<Int32> pAptQualifier)`, which isn't very helpful.)

To address this, I would like to project Win32 enums like `APTTYPE` and `APTTYPEQUALIFIER` as Dart extension types, as shown below:

```dart
extension type const APTTYPE(Int32 _) implements Int32 {
  static const APTTYPE_CURRENT = 0;
  static const APTTYPE_STA = 1;
  static const APTTYPE_MTA = 2;
  static const APTTYPE_NA = 3;
  static const APTTYPE_MAINSTA = 4;
}

extension type const APTTYPEQUALIFIER(Int32 _) implements Int32 {
  static const APTTYPEQUALIFIER_NONE = 0;
  static const APTTYPEQUALIFIER_IMPLICIT_MTA = 1;
  static const APTTYPEQUALIFIER_NA_ON_MTA = 2;
  static const APTTYPEQUALIFIER_NA_ON_STA = 3;
  static const APTTYPEQUALIFIER_NA_ON_IMPLICIT_MTA = 4;
  static const APTTYPEQUALIFIER_NA_ON_MAINSTA = 5;
  static const APTTYPEQUALIFIER_APPLICATION_STA = 6;
  static const APTTYPEQUALIFIER_RESERVED_1 = 7;
}
```

With this approach, users will see the exact enum types when hovering over the function, instead of seeing `Pointer<Int32>`, and they'll be able to allocate memory for the enums using `calloc`, as demonstrated below:
```dart
void main() {
  final pAptType = calloc<APTTYPE>();
  final pAptQualifier = calloc<APTTYPEQUALIFIER>();

  try {
    final hr = CoGetApartmentType(pAptType, pAptQualifier);
    if (hr == S_OK) {
      print('COM is initialized on the current thread.');

      if (pAptType.value == APTTYPE.APTTYPE_STA) {
        print('The current thread is in a single-threaded apartment.');
      } else if (pAptType.value == APTTYPE.APTTYPE_MTA) {
        print('The current thread is in a multi-threaded apartment.');
      }

      if (pAptQualifier.value == APTTYPEQUALIFIER.APTTYPEQUALIFIER_NONE) {
        print('No qualifier information for the current COM apartment type is '
            'available.');
      } else if (pAptQualifier.value ==
          APTTYPEQUALIFIER.APTTYPEQUALIFIER_IMPLICIT_MTA) {
        print('The current thread is in an implicit multi-threaded apartment.');
      }
    } else {
      print('COM is not initialized on the current thread.');
    }
  } finally {
    calloc
      ..free(pAptType)
      ..free(pAptQualifier);
  }
}
```

**Comment by dcharkes** [comment]
Thanks for the clear explanation! ❤️  (And thanks for continuing development on package:win32!)

So if I understand you correctly, you only want to use extension types as type markers in C function signatures and as type arguments to `Pointer` for type safety. Do you also have a need for passing values around or only types? (Well you cant have values of `Int32` anyway, but you could of structs.)

As for the question of marker types, I wonder if extension types are the right solution yes or no.

One alternative is to generate `class APTTYPE implements AbiSpecificInteger`. This will also work for if you have `APTTYPE` directly in function signatures.

Both of these solutions will work today. Though, there might be downsides to doing this that I haven't thought of.

WDYT? Any thoughts, ideas?

(Side note: We have `class Foo extends Opaque` as a marker type. If we go the direction of extension types, we should probably have people migrate to having an extension type on opaque, rather than extending it. Because you want to use the integer-size with `calloc` you can't use `Opaque` for your use case.)

**Comment by halildurmus** [comment]
> Thanks for the clear explanation! ❤️ (And thanks for continuing development on package:win32!)

Thanks!

> So if I understand you correctly, you only want to use extension types as type markers in C function signatures and as type arguments to `Pointer` for type safety. 

Exactly.

> Do you also have a need for passing values around or only types? (Well you cant have values of `Int32` anyway, but you could of structs.)

Yes. Some Win32 APIs directly accept native integer types and structs by value. As you said, given that we can't create values of `Int32` anyway, I'll keep using the Dart equivalents (e.g., `int`, `double`)  in those cases -- and that's fine. But I could leverage extension types for APIs that take structs by value though.

> As for the question of marker types, I wonder if extension types are the right solution yes or no.
> 
> One alternative is to generate `class APTTYPE implements AbiSpecificInteger`. This will also work for if you have `APTTYPE` directly in function signatures.

I couldn't understand this option. Let's say the `base` modifier on AbiSpecificInteger is removed and one can declare a class like `class APTTYPE implements AbiSpecificInteger`. How would this function exactly? How do I specify which native type (Int32, Uint32) this class represents so that when I allocate memory with `calloc<APTTYPE>()`, the size of that native type is used?

**Comment by dcharkes** [comment]
I meant to say `extends` not `implements`, my bad! See the documentation here: https://api.dart.dev/stable/3.3.0/dart-ffi/AbiSpecificInteger-class.html You'd just add the same integer size for all windows targets:

```dart
@AbiSpecificIntegerMapping({
  Abi.windowsArm64: Int32(),
  Abi.windowsIA32: Int32(),
  Abi.windowsX64: Int32(),
})
final class APTTYPE extends AbiSpecificInteger {
  const APTTYPE();
}
```

**Comment by halildurmus** [comment]
That's nice, thanks for the clarification.

However, there is an important downside to using this option. We can't seem to define members for this class 😢:
![Screenshot 2024-02-16 212604](https://github.com/dart-lang/sdk/assets/45524029/725974fb-c407-4f01-8d75-057c21a06468)

Can this restriction be removed?

**Comment by dcharkes** [comment]
Right, static members should be fine. I can make a fix for that.

Are all the types that you want to wrap with documentation integer types?
Do you have any structs that are identical but need a different documentation? (You could generate two identical structs in that case.)

**Comment by dcharkes** [comment]
> There is an important difference between PWSTR and BSTR types. BSTR has a four-byte integer prefix that contains the number of bytes in the following data string. This means one can't use the `toNativeUtf16()` method to create a BSTR type. Instead, they need to use the [SysAllocString](https://learn.microsoft.com/en-us/windows/win32/api/oleauto/nf-oleauto-sysallocstring) function to create a BSTR. That's why I think it is important that I wrap these types as well. However, this does not seem possible with the AbiSpecificInteger.

You could define a new `Opaque` type for those, and not relate it at all to `Utf16`. Since they are not compatible with the `Pointer<Utf16>` at all, your users should never get the base-type out of the extension type.

Your API would be centered around `Pointer<PWSTR>` and have extension methods on that type (and on `String`) rather than `PWSTR` itself representing a `Pointer<something>`. This is also how the `Pointer<Utf16>` API is structured.

Would this work?

I kind of like the extension types as a general wrapper method, because if we allow it, it would work for any native type anywhere in the FFI.

I am slightly worried that it might conflict with us wanting to migrate the `dart:ffi` API to extension types (for example `Int8` being an extension type on `int`).

* https://github.com/dart-lang/sdk/issues/54504

@lrhn @mkustermann Any thoughts about allowing users to wrap any `NativeType` in extension types and automatically unwrapping/wrapping?

* Automatic unwrapping on FFI call arguments / FFI callback return value / Pointer store operations
* Automatic wrapping on FFI call return value / FFI callback arguments / Pointer load operations

The only use case I can think of is custom code generators. Users using `package:ffigen` will not use this feature. Unless we start adding logic to FFIgen to target these. We have a work-around in FFIgen to match certain types and generate another type that is being used to turn `char*` into `Pointer<Utf8>`. If we introduce extension type wrapping in `dart:ffi`, we should probably generalize that mechanism in `package:ffi`. Issue tracking this for strings:

* https://github.com/dart-lang/native/issues/315

**Comment by eernstg** [comment]
Interesting!!

I'm just wondering, would these ideas have some synergy (or the opposite) with https://github.com/dart-lang/language/issues/3614? That's a proposal to allow some extension types to be assignable from their representation type:

```dart
extension type Int32(int _) implements int {}
extension type AptType(Int32 _) implements Int32 {}

Int32 i32 = 42; // OK.
AptType aptType = i32; // Could be OK if the proposal is generalized slightly.
AptType aptType2 = 42; // Perhaps --- that would be yet another generalization.
```

> I think making the FFI static analysis / FFI kernel transformer simply unwrap all extension types to their representation types before doing the checking would seem fine to me.

To me this sounds like it could erase some typing distinctions that we'd want to maintain. Would it be sufficient to erase the extension type when it's used to select a C type?

**Comment by lrhn** [comment]
The original request looks like a non-issue to allow.

If you do `extension type MyInt32(Int32 _) implements Int32 { ... }`, then there is no doubt that you're just adding members to `Int32`, or removing some. The operative type is still clearly `Int32`.

If you do `extension type MyInt32(Int32 _) { ... }` with no supertype, then I wouldn't let the type count as `Int32`.

If you do `extension type MyInt32(Int32 _) implements NativeType { ... }`... I'd still probably not let it work like `Int32`, and then it's limited where it can be used.

So, to allow it, the extension type must implement a concrete native type that could be used somewhere, and must also have a representation type that is a subtype of that type. I don't think there are many useful type hierarchies where among native types, so that probably just means implementing the representation type.
(You can also choose to make that the criterion for when you can use an extension type as a type argument where you'd normally require a native type.)


Wrapping pointers is probably where the fun is.
```dart
extension type BStr(Pointer<Uint16> _) implements Pointer<Uint16> {
  int get length => _.cast<Uint32>().value;
  Uint16 operator[](int index) => _[2 + IndexError.check(index, length)];
}
```
I'm not sure I'd make that type *implement* `Pointer<Uint16>` though. It feels like an abstraction that would be better served by having a `Pointer<Uint16> get address` than *being* the pointer. (Favor composition over inheritance.)

(Can you wrap an `Array<Uint16>`, which means creating an `Array<Uint16>` value, or is that only a marker type?
In C, I'd make it a struct like `struct {uint32_t length, uint16_t[0] chars}`, then access `theStruct.chars[i]`, but not sure is a good idea for FFI, since it doesn't actually have a size.)

About making the native types themselves be extensions, that's precisely the kind of thing extension types are made for, so it's no surprise it seems like a good fit. The places where it might fall short is that extension types are not good at being *abstract supertypes*. I'd probably use real classes for the `Uint32`-like "not real types", and extension types for the types that users can interact with.
```dart
sealed class NativeType {}
sealed class NativeIntegerType extends NativeType< {}
final class Uint32 extends NativeIntegerType {
  static const int size = 4;
}
final class Uint16 extends NativeIntegerType {}
// ...

extension type Pointer<T extends NativeType>(int address) {
   Pointer<R> cast<R extends NativeType>() => Pointer<R>(address);
   bool operator <(Pointer<NativeType> other) => address < other.address;
   bool operator <=(Pointer<NativeType> other) => address <= other.address;
   bool operator >(Pointer<NativeType> other) => address > other.address;
   bool operator >=(Pointer<NativeType> other) => address >= other.address;
}
extension Uint32Pointer on Pointer<Uint32> {
  int get value => _readUint32(address);
  set value(int value) { _writeUint32(address, value); }
  Pointer<Uint32> operator +(int offset) => Pointer<Uint32>(address + offset * Uint32.size);
  Pointer<Uint32> operator -(int offset) => Pointer<Uint32>(address - offset * Uint32.size);
  int distance(Pointer<Uint32> other) => (other.address - this.address) ~/ Uint32.size;
  int operator [](int offset) => (this + offset).value;
  operator []=(int offset, int value) { (this + offset).value = value; }
}
```
(The `Pointer` class has very little functionality itself. Which is not surprising, there is little you can do with a `C` pointer if you don't know its type, so nothing you can't do with a NULL void pointer. But `==` works.)

**Comment by halildurmus** [comment]
> Wrapping pointers is probably where the fun is.
> 
> ```dart
> extension type BStr(Pointer<Uint16> _) implements Pointer<Uint16> {
>   int get length => _.cast<Uint32>().value;
>   Uint16 operator[](int index) => _[2 + IndexError.check(index, length)];
> }
> ```

Just to clarify, the pointer for the [BSTR](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/automat/bstr) type points to the first character of the data string, not to the length prefix.

So your example should be modified as follows:
```dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:win32/win32.dart';

extension type BStr(Pointer<Uint16> _) implements Pointer<Uint16> {
  int get length =>
      Pointer<Uint32>.fromAddress(address - sizeOf<Uint32>()).value;
}

void main() {
  final bstr = BStr(SysAllocString('Hello, world!'.toNativeUtf16()).cast());
  print(bstr.length); // 26 = 13 * sizeOf<Uint16>()
  print(bstr[0]); // 72 (H)
  print(bstr[12]); // 33 (!)
}
```

Also, Win32 API provides [String Manipulation Functions](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/automat/string-manipulation-functions) for working with BSTR types.

Here's a simple example that uses these APIs:
```dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:win32/win32.dart';

extension type BStr(Pointer<Uint16> _) implements Pointer<Uint16> {
  factory BStr.fromString(String string) {
    final psz = string.toNativeUtf16();
    final pbstr = SysAllocString(psz);
    calloc.free(psz);
    return BStr(pbstr.cast());
  }

  int get length => SysStringLen(this.cast());

  void free() => SysFreeString(this.cast());
}
```
